bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/HueBridgeHandler.java

```diff
@@ -178,44 +173,55 @@ private boolean isReachable(String ipAddress) {
-                    final Map<String, Object> lastFullSensorState = lastFullSensor.getState();
-                    lastSensorStates.put(sensorId, sensor);
-                    if (!lastFullSensorState.equals(sensor.getState())) {
-                        logger.debug("Status update for Hue sensor '{}' detected: {}", sensorId, sensor.getState());
-                        notifySensorStatusListeners(sensor, STATE_CHANGED);
-                    }
-                } else {
-                    lastSensorStates.put(sensorId, sensor);
+                lastSensorStateCopy.remove(sensorId);
+
+                final SensorStatusListener sensorStatusListener = sensorStatusListeners.get(sensorId);
+                if (sensorStatusListener == null) {
                     logger.debug("Hue sensor '{}' added.", sensorId);
-                    notifySensorStatusListeners(sensor, STATE_ADDED);
 
+                    if (discovery != null) {
+                        discovery.onSensorAdded(hueBridge, sensor);
+                    }
+
+                    lastSensorStates.put(sensorId, sensor);
+                } else {
+                    lastSensorStateCopy.remove(sensorId);
+                    if (sensorStatusListener.onSensorStateChanged(hueBridge, sensor)) {
+                        lastSensorStates.put(sensorId, sensor);
+                    }
                 }
             }
 
             // Check for removed sensors
-            for (Entry<String, FullSensor> fullSensorEntry : lastSensorStateCopy.entrySet()) {
-                lastSensorStates.remove(fullSensorEntry.getKey());
-                logger.debug("Hue sensor '{}' removed.", fullSensorEntry.getKey());
-                for (SensorStatusListener sensorStatusListener : sensorStatusListeners) {
+            lastSensorStateCopy.forEach((sensorId, sensor) -> {
+                logger.debug("Hue sensor '{}' removed.", sensorId);
+                final SensorStatusListener sensorStatusListener = sensorStatusListeners.get(sensorId);
+                if (sensorStatusListener != null) {
                     try {
-                        sensorStatusListener.onSensorRemoved(hueBridge, fullSensorEntry.getValue());
+                        sensorStatusListener.onSensorRemoved(hueBridge, sensor);
                     } catch (Exception e) {
                         logger.error("An exception occurred while calling the Sensor Listeners", e);
                     }
                 }
-            }
+
+                if (discovery != null) {
+                    discovery.onSensorRemoved(hueBridge, sensor);
+                }
+            });
         }
     };
 
     private final Runnable lightPollingRunnable = new PollingRunnable() {
         @Override
         protected void doConnectedRun() throws IOException, ApiException {
             Map<String, FullLight> lastLightStateCopy = new HashMap<>(lastLightStates);
+            lastLightStates.clear();
```

Comment:
If you clear all the states, you will loose the previous state in case it is not changed by the thing handler.

Reply:
Changed this!